test/throughput/test_batch_profile.py

```diff
@@ -0,0 +1,120 @@
+def data():
+    print("Initialising data...")
+    np.random.seed(0)
+    batch_set = [Batch(a=[j for j in np.arange(1e3)],
+                       b={'b1': (3.14, 3.14), 'b2': np.arange(1e3)},
+                       c=i) for i in np.arange(int(1e4))]
+    batch0 = Batch(
+        a=np.ones((3, 4), dtype=np.float64),
+        b=Batch(
+            c=np.ones((1,), dtype=np.float64),
+            d=torch.ones((3, 3, 3), dtype=torch.float32),
+            e=list(range(3))
+        )
+    )
+    batch1 = copy.deepcopy(batch0)
+    batch2 = copy.deepcopy(batch0)
+    batch_len = int(1e4)
+    batch3 = Batch(obs=[np.arange(20) for _ in np.arange(batch_len)],
+                   reward=np.arange(batch_len))
+    indexs = np.random.choice(batch_len,
+                              size=batch_len//10, replace=False)
+    slice_dict = {'obs': [np.arange(20)
+                          for _ in np.arange(batch_len//10)],
+                  'reward': np.arange(batch_len//10)}
+    dict_set = [{'obs': np.arange(20), 'info': "this is info", 'reward': 0}
+                for _ in np.arange(1e2)]
+    batch4 = Batch(
+        a=np.ones((10000, 4), dtype=np.float64),
+        b=Batch(
+            c=np.ones((1,), dtype=np.float64),
+            d=torch.ones((1000, 1000), dtype=torch.float32),
+            e=np.arange(1000)
+        )
+    )
+
+    print("Initialised")
+    return {'batch_set': batch_set,
+            'batch0': batch0,
+            'batch1': batch1,
+            'batch2': batch2,
+            'batch3': batch3,
+            'indexs': indexs,
+            'dict_set': dict_set,
+            'slice_dict': slice_dict,
+            'batch4': batch4
+            }
+
+
+def test_init(data):
+    """Test Batch __init__()."""
+    for _ in np.arange(10):
+        _ = Batch(data['batch_set'])
+
+
+def test_get_item(data):
+    """Test get with item."""
+    for _ in np.arange(1e5):
+        _ = data['batch3'][data['indexs']]
+
+
+def test_get_attr(data):
+    """Test get with attr."""
+    for _ in np.arange(1e6):
+        data['batch3'].get('obs')
+        data['batch3'].get('reward')
+        _, _ = data['batch3'].obs, data['batch3'].reward
+
+
+def test_set_item(data):
+    """Test set with item."""
+    for _ in np.arange(1e4):
+        data['batch3'][data['indexs']] = data['slice_dict']
+
+
+def test_set_attr(data):
+    """Test set with attr."""
+    for _ in np.arange(1e4):
+        data['batch3'].c = np.arange(1e3)
+        data['batch3'].obs = data['dict_set']
+
+
+def test_numpy_torch_convert(data):
+    """Test conversion between numpy and torch."""
+    for _ in np.arange(1e5):
+        data['batch4'].to_torch()
+        data['batch4'].to_numpy()
+
+
+def test_pickle(data):
+    for _ in np.arange(1e4):
+        pickle.loads(pickle.dumps(data['batch4']))
+
+
+def test_cat(data):
+    """Test cat"""
+    for _ in np.arange(1e4):
+        Batch.cat((data['batch0'], data['batch0']))
+        data['batch1'].cat_(data['batch0'])
```

Comment:
This seems strange to me since cat_ is a in-place operation and batch1 will increase the length in every iteration.

Reply:
yes? I intend to increase the length of batch every iteration, because batch1 is only used for in place cat_(). How do you purpose we change this ?

Reply:
Now I understand why it is weird, fixed.